File: Animator/app/src/main/java/com/maxxipoint/animator/LikeLayout.java
//package com.maxxipoint.animator;
//
//import android.animation.Animator;
//import android.animation.AnimatorListenerAdapter;
//import android.animation.AnimatorSet;
//import android.animation.ObjectAnimator;
//import android.animation.TypeEvaluator;
//import android.animation.ValueAnimator;
//import android.content.Context;
//import android.graphics.Matrix;
//import android.graphics.PointF;
//import android.graphics.drawable.Drawable;
//import android.util.AttributeSet;
//import android.util.Log;
//import android.view.LayoutInflater;
//import android.view.MotionEvent;
//import android.view.View;
//import android.view.animation.LinearInterpolator;
//import android.widget.FrameLayout;
//import android.widget.ImageView;
//import android.widget.RelativeLayout;
//
//import androidx.annotation.NonNull;
//import androidx.annotation.Nullable;
//
//import java.util.Random;
//
//public class LikeLayout extends FrameLayout {
//
//    ImageView imageView,top,left,right,leftBottom,rightBottom;
//
//    int window_width = 300; // 屏幕宽度
//    int window_height = 300 ; // 屏幕高度
//    int image_width = 60; // 小红心心的宽度
//
//    Drawable icon = getResources().getDrawable(R.mipmap.ic_heart);
//    int mClickCount = 0;     //点击一次是暂停，多次是点赞
//
//    public LikeLayout(@NonNull Context context) {
//        super(context);
//        initView();
//    }
//
//    public LikeLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
//        super(context, attrs);
//        initView();
//    }
//
//    public LikeLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
//        super(context, attrs, defStyleAttr);
//        initView();
//    }
//
//    private void initView() {
//        setClipChildren(false);  // 避免旋转时红心被遮挡
//
//        View viewRoot = LayoutInflater.from(getContext()).inflate(R.layout.animator_layout,this,true);
//        imageView = viewRoot.findViewById(R.id.icon);
//        top = viewRoot.findViewById(R.id.shop); // 顶部的图片控件
//        left = viewRoot.findViewById(R.id.news); // 左边的图片控件
//        right = viewRoot.findViewById(R.id.fans); // 右边的图片控件
//        leftBottom = viewRoot.findViewById(R.id.trails); // 左下角图片控件
//        rightBottom = viewRoot.findViewById(R.id.expression); // 右下角图片控件
//    }
//
//    @Override
//    public boolean onTouchEvent(MotionEvent event) {
//        if (event.getAction() == MotionEvent.ACTION_DOWN) {     // 按下时在Layout中生成红心
//            float x = event.getX();
//            float y = event.getY();
//            mClickCount++;
////            mHandler.removeCallbacksAndMessages(null);
//            if (mClickCount >= 2) { // 双击事件
//                addHeartView(x, y);  // 在Layout中添加红心并，播放消失动画
////                onLikeListener();
////                mHandler.sendEmptyMessageDelayed(1, 500); // 清除所有的 handler 消息
//            } else { // 单击事件
////                mHandler.sendEmptyMessageDelayed(0, 500);
//            }
//
//        }
//        return true;
//    }
//
//    /**
//     * 在Layout中添加红心并，播放消失动画
//     */
//    private void addHeartView(float x, float y) {
//        //计算点击的点位红心的下部中间
//        LayoutParams lp = new LayoutParams(icon.getIntrinsicWidth(), icon.getIntrinsicHeight());
//
//        lp.leftMargin = (int) (x - icon.getIntrinsicWidth() / 2);
//        lp.topMargin = (int) (y - icon.getIntrinsicHeight());
//
//        Log.e("点击左上角X坐标 = ","${lp.leftMargin}");
//        Log.e("点击左上角Y坐标 = ","${lp.topMargin}");
//        Log.e("点击图片的宽 = ","${icon.intrinsicWidth}");
//        Log.e("点击图片的高 = ","${icon.intrinsicHeight}");
//        window_height = Math.abs(lp.topMargin) + icon.getIntrinsicHeight() * 4;
//        window_width = Math.abs(lp.leftMargin) + icon.getIntrinsicWidth() * 4;
//
//        Log.e("window_width = ","$window_width");
//        Log.e("window_height = ","$window_height");
//
////        val imageView = ImageView(context)
//        imageView.setScaleType(ImageView.ScaleType.MATRIX);  // 使用图像矩阵缩放
//        Matrix matrix = new Matrix();
//        matrix.postRotate(getRandomRotate());      //设置红心的微量偏移
//        imageView.setImageMatrix(matrix);   // 给 ImageView 设置矩阵
//        imageView.setImageDrawable(icon);  // 给 ImageView 设置展示的图片
//        imageView.setLayoutParams(lp);
//
//        // 这部分代码可使 下层的 5个图片控件跟着心形的坐标走
//        LayoutParams lpView = new LayoutParams(0, 0);   //计算点击的点位红心的下部中间
//        lpView.leftMargin = (int) (x - icon.getIntrinsicWidth() / 2);
//        lpView.topMargin = (int) (y - icon.getIntrinsicHeight());
//        top.setLayoutParams(lp);
//        left.setLayoutParams(lp);
//        right.setLayoutParams(lp);
//        leftBottom.setLayoutParams(lp);
//        rightBottom.setLayoutParams(lp);
////        addView(imageView)
//
//        AnimatorSet animSet = getShowAnimSet(imageView);
//        final AnimatorSet hideSet = getHideAnimSet(imageView);
//        animSet.start();
//        animSet.addListener(new AnimatorListenerAdapter() {
//            @Override
//            public void onAnimationEnd(Animator animation) {
//                super.onAnimationEnd(animation);
//                hideSet.start();
//            }
//        });
//        hideSet.addListener(new AnimatorListenerAdapter() {
//            @Override
//            public void onAnimationEnd(Animator animation) {
//                super.onAnimationEnd(animation);
//                top.setAlpha(0f);
//                left.setAlpha(0f);
//                right.setAlpha(0f);
//                leftBottom.setAlpha(0f);
//                rightBottom.setAlpha(0f);
//            }
//        });
//    }
//
//    /**
//     * 刚点击的时候的一个缩放效果
//     */
//    private AnimatorSet getShowAnimSet(ImageView view) {
//        // 缩放动画
//        ObjectAnimator scaleX = ObjectAnimator.ofFloat(view, "scaleX", 1.2f, 1f);
//        ObjectAnimator scaleY = ObjectAnimator.ofFloat(view, "scaleY", 1.2f, 1f);
//        AnimatorSet animSet = new AnimatorSet();
//        animSet.playTogether(scaleX, scaleY);
//        animSet.setDuration(300);
//        return animSet;
//    }
//
//    /**
//     * 缩放结束后到红心消失的效果
//     */
//    private AnimatorSet getHideAnimSet(ImageView view) {
//        // 1.alpha动画
//        ObjectAnimator alpha = ObjectAnimator.ofFloat(view, "alpha", 1f, 0.0f);
//        // 2.缩放动画
//        ObjectAnimator scaleX = ObjectAnimator.ofFloat(view, "scaleX", 1f, 2f);
//        ObjectAnimator scaleY = ObjectAnimator.ofFloat(view, "scaleY", 1f, 2f);
//        // 3.translation动画
//        ObjectAnimator translation = ObjectAnimator.ofFloat(view, "translationY", 0f, -150f);
//
//        // 底下动画开始
//        top.setAlpha(1f);
//        left.setAlpha(1f);
//        right.setAlpha(1f);
//        leftBottom.setAlpha(1f);
//        rightBottom.setAlpha(1f);
//
//        int startangle = 54; // 角度 方向 +1 正方向，-1 负方向。
//
//        // Math.PI 圆周率 3.141592653589793
//        // 圆周率乘以72°再除以180°是:  Math.sin(72 * Math.PI / 180
//        // 是72角对应的弧度值
//
//        int length = (int) ((window_width
//                / (Math.sin(72 * Math.PI / 180)) - image_width) / 2 - image_width / 5);
//
//        // 右下角图片设置动画
//        ValueAnimator expressionanim = zhixian(rightBottom, startangle,
//                length);
//        // 左下角图片设置动画
//        ValueAnimator trailsanim = zhixian(leftBottom, startangle + 72,
//                length);
//        // 左边图片设置动画
//        ValueAnimator newsanim = zhixian(left, startangle + (72 * 2),
//                length);
//        // 顶部图片设置动画
//        ValueAnimator shopanim = zhixian(top, startangle + (72 * 3),
//                length);
//        // 右边图片设置动画
//        ValueAnimator fansanim = zhixian(right, startangle + (72 * 4),
//                length);
//
//        // 底下动画 end
//
//        AnimatorSet animSet = new AnimatorSet();
//        animSet.playTogether(alpha, scaleX, scaleY, translation,shopanim, fansanim,
//                newsanim, expressionanim, trailsanim);
//        animSet.setDuration(500);
//        return animSet;
//    }
//
//    /**
//     * 沿直线运动。
//     *
//     * @param view  要移动的对象。
//     * @param angle  方向 +1 正方向，-1 负方向。
//     * @param length 移动的距离
//     */
//    public ValueAnimator zhixian(final View view, final int angle, final int length) {
//        ValueAnimator valueAnimator = new ValueAnimator();
//        valueAnimator.setObjectValues(new PointF(0, 0));
//        valueAnimator.setInterpolator(new LinearInterpolator());
//        valueAnimator.setEvaluator(new TypeEvaluator<PointF>() {
//            // fraction = t / duration
//            @Override
//            public PointF evaluate(float fraction, PointF startValue,
//                                   PointF endValue) {
//                Log.v("znz", "znz ---> " + fraction);
//                PointF point = new PointF();
//                point.x = (float) (fraction * length * Math.cos(angle * Math.PI
//                        / 180));
//                point.y = (float) (fraction * length * Math.sin(angle * Math.PI
//                        / 180));
//                return point;
//            }
//        });
//        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
//            @Override
//            public void onAnimationUpdate(ValueAnimator animation) {
//                PointF point = (PointF) animation.getAnimatedValue();
//                FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
//                        image_width, image_width);
//                params.leftMargin = (int) point.x
//                        + (window_width / 2 - image_width / 2); // Your
//                // coordinate
//                params.topMargin = (int) point.y
//                        + (window_height / 2 - image_width / 2); // Your Y
//                // coordinate
//                Log.v("znz", "point.x ---> " + point.x);
//                Log.v("znz", "point.y ---> " + point.y);
//                view.setLayoutParams(params);
//            }
//        });
//        return valueAnimator;
//    }
//
//    /**
//     * 生成一个随机的左右偏移量
//     */
//    private float getRandomRotate() {
//        return (float) (new Random().nextInt(40) - 10);
//    }
//}

File: Animator/app/src/main/java/com/maxxipoint/animator/LikeLayout.kt
package com.maxxipoint.animator

import android.animation.*
import android.content.Context
import android.graphics.Matrix
import android.graphics.PointF
import android.graphics.drawable.Drawable
import android.os.Handler
import android.os.Message
import android.util.AttributeSet
import android.util.Log
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.widget.FrameLayout
import android.widget.ImageView
import com.maxxipoint.animator.LikeLayout
import java.lang.ref.WeakReference
import java.util.*


class LikeLayout : FrameLayout {

    var onLikeListener: () -> Unit = {}     //屏幕点赞后，点赞按钮需同步点赞
    var onPauseListener: () -> Unit = {}     //暂停 或者 继续播放

    lateinit var imageView: ImageView
    lateinit var top: ImageView
    lateinit var left: ImageView
    lateinit var right: ImageView
    lateinit var leftBottom: ImageView
    lateinit var rightBottom: ImageView

    var window_width = 0 // 屏幕宽度
    var window_height = 0  // 屏幕高度
    val image_width = 60 // 小红心心的宽度

    constructor(context: Context) : super(context){
        initView()
    }
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        initView()
    }
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr){
        initView()
    }

    private var icon: Drawable = resources.getDrawable(R.mipmap.ic_heart)
    private var mClickCount = 0     //点击一次是暂停，多次是点赞
    private val mHandler = LikeLayoutHandler(this)

    init {
        clipChildren = false        //避免旋转时红心被遮挡
    }

    private fun initView() {
        val viewRoot = LayoutInflater.from(context).inflate(R.layout.animator_layout,this,true)
        imageView = viewRoot.findViewById(R.id.icon)
        top = findViewById(R.id.shop) // 顶部的图片控件
        left = findViewById(R.id.news) // 左边的图片控件
        right = findViewById(R.id.fans) // 右边的图片控件
        leftBottom = findViewById(R.id.trails) // 左下角图片控件
        rightBottom = findViewById(R.id.expression) // 右下角图片控件
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        if (event?.action == MotionEvent.ACTION_DOWN) {     // 按下时在Layout中生成红心
            val x = event.x
            val y = event.y
            mClickCount++
            mHandler.removeCallbacksAndMessages(null)
            if (mClickCount >= 2) { // 双击事件
                addHeartView(x, y)  // 在Layout中添加红心并，播放消失动画
                onLikeListener()
                mHandler.sendEmptyMessageDelayed(1, 500)
            } else { // 单击事件
                mHandler.sendEmptyMessageDelayed(0, 500)
            }

        }
        return true
    }

    // 暂停播放事件
    private fun pauseClick() {
        if (mClickCount == 1) {
            onPauseListener()
        }
        mClickCount = 0
    }

    // 在 activity 或者 fragment 的 onPause() 方法中执行
    fun onPause() {
        mClickCount = 0
        mHandler.removeCallbacksAndMessages(null)
    }

    lateinit var lp: LayoutParams
    /**
     * 在Layout中添加红心并，播放消失动画
     */
    private fun addHeartView(x: Float, y: Float) {

        lp = LayoutParams(icon.intrinsicWidth, icon.intrinsicHeight)    //计算点击的点位红心的下部中间

        lp.leftMargin = (x - icon.intrinsicWidth / 2).toInt()
        lp.topMargin = (y - icon.intrinsicHeight/2).toInt()

        Log.e("点击左上角X坐标 = ","${lp.leftMargin}")
        Log.e("点击左上角Y坐标 = ","${lp.topMargin}")
        Log.e("点击图片的宽 = ","${icon.intrinsicWidth}")
        Log.e("点击图片的高 = ","${icon.intrinsicHeight}")
        window_height = y.toInt();//Math.abs(lp.topMargin) + icon.intrinsicHeight
        window_width =x.toInt();// Math.abs(lp.leftMargin) + icon.intrinsicWidth

//        val imageView = ImageView(context)
        imageView.scaleType = ImageView.ScaleType.MATRIX  // 使用图像矩阵缩放
        val matrix = Matrix()
        matrix.postRotate(getRandomRotate())       //设置红心的微量偏移
        imageView.imageMatrix = matrix   // 给 ImageView 设置矩阵
        imageView.setImageDrawable(icon)  // 给 ImageView 设置展示的图片
        imageView.layoutParams = lp

        // 这部分代码可使 下层的 5个图片控件跟着心形的坐标走
        val lpView = LayoutParams(0, 0)    //计算点击的点位红心的下部中间
        lpView.leftMargin = (x - image_width / 2).toInt()
        lpView.topMargin = (y - image_width/2).toInt()
        top.layoutParams = lpView
        left.layoutParams = lpView
        right.layoutParams = lpView
        leftBottom.layoutParams = lpView
        rightBottom.layoutParams = lpView

        top.alpha = 1f
        left.alpha = 1f
        right.alpha = 1f
        leftBottom.alpha = 1f
        rightBottom.alpha = 1f
//        addView(imageView)

        val animSet = getShowAnimSet(imageView)
        val hideSet = getHideAnimSet(imageView)
        animSet.start()
        animSet.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator?) {
                super.onAnimationEnd(animation)
                hideSet.start()
            }
        })
        hideSet.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator?) {
                super.onAnimationEnd(animation)
//                removeView    //动画结束移除红心
//                imageView.visibility = View.GONE
//
                top.alpha = 0f
                left.alpha = 0f
                right.alpha = 0f
                leftBottom.alpha = 0f
                rightBottom.alpha = 0f
            }
        })
    }

    /**
     * 刚点击的时候的一个缩放效果
     */
    private fun getShowAnimSet(view: ImageView): AnimatorSet {
        // 缩放动画
        val scaleX = ObjectAnimator.ofFloat(view, "scaleX", 1.2f, 1f)
        val scaleY = ObjectAnimator.ofFloat(view, "scaleY", 1.2f, 1f)
        val animSet = AnimatorSet()
        animSet.playTogether(scaleX, scaleY)
        animSet.duration = 300
        return animSet
    }

    /**
     * 缩放结束后到红心消失的效果
     */
    private fun getHideAnimSet(view: ImageView): AnimatorSet {
        // 1.alpha动画
        val alpha = ObjectAnimator.ofFloat(view, "alpha", 1f, 0.0f)
        // 2.缩放动画
        val scaleX = ObjectAnimator.ofFloat(view, "scaleX", 1f, 2f)
        val scaleY = ObjectAnimator.ofFloat(view, "scaleY", 1f, 2f)
        // 3.translation动画
        val translation = ObjectAnimator.ofFloat(view, "translationY", 0f, -150f)

        // 底下动画开始
        top.alpha = 1f
        left.alpha = 1f
        right.alpha = 1f
        leftBottom.alpha = 1f
        rightBottom.alpha = 1f

        val startangle = 54 // 角度 方向 +1 正方向，-1 负方向。

        // Math.PI 圆周率 3.141592653589793
        // 圆周率乘以72°再除以180°是:  Math.sin(72 * Math.PI / 180
        // 是72角对应的弧度值
        val length = ((600
                / Math.sin(72 * Math.PI / 180) - image_width) / 2 - image_width / 5).toInt()

        // 右下角图片设置动画
        val translationRightBottom = zhixian(
            rightBottom, startangle,
            length
        )
        val alphaRightBottom = ObjectAnimator.ofFloat(rightBottom, "alpha", 1f, 0.0f)
        // 左下角图片设置动画
        val translationLeftBottom = zhixian(
            leftBottom, startangle + 72,
            length
        )
        val alphaLeftBottom = ObjectAnimator.ofFloat(leftBottom, "alpha", 1f, 0.0f)
        // 左边图片设置动画
        val translationLeft = zhixian(
            left, startangle + 72 * 2,
            length
        )
        val alphaLeft = ObjectAnimator.ofFloat(left, "alpha", 1f, 0.0f)
        // 顶部图片设置动画
        val translationTop = zhixian(
            top, startangle + 72 * 3,
            length
        )
        val alphaTop = ObjectAnimator.ofFloat(top, "alpha", 1f, 0.0f)
        // 右边图片设置动画
        val translationRight = zhixian(
            right, startangle + 72 * 4,
            length
        )
        val alphaRight = ObjectAnimator.ofFloat(right, "alpha", 1f, 0.0f)
        // 底下动画 end

        val animSet = AnimatorSet()
        animSet.playTogether(alpha, scaleX, scaleY, translation,alphaTop,translationTop, alphaRight,translationRight,
            alphaLeft,translationLeft, alphaRightBottom, translationRightBottom, alphaLeftBottom,translationLeftBottom)
        animSet.duration = 500
        return animSet
    }

    /**
     * 生成一个随机的左右偏移量
     */
    private fun getRandomRotate(): Float = (Random().nextInt(40) - 10).toFloat()

    companion object {
        private class LikeLayoutHandler(view: LikeLayout) : Handler() {
            private val mView = WeakReference(view)

            override fun handleMessage(msg: Message) {
                super.handleMessage(msg)
                when(msg?.what){
                    0-> mView.get()?.pauseClick()
                    1-> mView.get()?.onPause()
                }
            }
        }
    }

    /**
     * 沿直线运动。
     *
     * @param view  要移动的对象。
     * @param angle  方向 +1 正方向，-1 负方向。
     * @param length 移动的距离
     */
    fun zhixian(view: View, angle: Int, length: Int): ValueAnimator? {
        val valueAnimator = ValueAnimator()
        // 设置Object型变化值，一般设置初始值和结束值，当然你也可以设置中间值，因为这是一个可变参数，长度可变
        valueAnimator.setObjectValues(PointF(0f, 0f))
//        valueAnimator.interpolator = LinearInterpolator()
        valueAnimator.interpolator = AccelerateInterpolator()

        // 设置动画的估值器
        valueAnimator.setEvaluator { fraction, startValue, endValue ->

            // fraction = t / duration
            Log.v("znz", "znz ---> $fraction")
            val point = PointF()
            point.x = (fraction * length * Math.cos(
                angle * Math.PI
                        / 180
            )).toFloat()
            point.y = (fraction * length * Math.sin(
                angle * Math.PI
                        / 180
            )).toFloat()
            point
        }
        valueAnimator.addUpdateListener { animation ->

            /**
             * 通过这样一个监听事件，我们就可以获取
             * 到ValueAnimator每一步所产生的值。
             *
             * 通过调用getAnimatedValue()获取到每个时间因子所产生的Value。
             */
            val point = animation.animatedValue as PointF // 获取到每个时间因子所产生的Value
            val params = FrameLayout.LayoutParams(image_width, image_width)
            params.leftMargin = (point.x.toInt()
                    + (window_width  - image_width / 2))  //  X 坐标
            params.topMargin = (point.y.toInt()
                    + (window_height  - image_width / 2)) //  Y 坐标
            Log.e("znz", "point.x ---> " + point.x)
            Log.e("znz", "point.y ---> " + point.y)
            view.layoutParams = params
        }
        return valueAnimator
    }
}